File: 104. Maximum Depth of Binary Tree/BinaryTreeMaxDepth.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # Base case: empty tree has depth 0
        if not root:
            return 0
        # Recursively compute depths of left and right subtrees
        left_depth = self.maxDepth(root.left)
        right_depth = self.maxDepth(root.right)
        # Return depth of current node: 1 + max of left and right depths
        return 1 + max(left_depth, right_depth)

File: 104. Maximum Depth of Binary Tree/BinaryTreeMaxDepth.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * Calculates the maximum depth of a binary tree.
     * @param root Root node of the binary tree
     * @return The maximum depth (number of nodes along the longest path to a leaf)
     */
    public int maxDepth(TreeNode root) {
        // Base case: empty tree has depth 0
        if (root == null) {
            return 0;
        }
        // Recursively compute depths of left and right subtrees
        int leftDepth = maxDepth(root.left);
        int rightDepth = maxDepth(root.right);
        // Return depth of current node: 1 + max of left and right depths
        return 1 + Math.max(leftDepth, rightDepth);
    }
}